keps/core/model-cross-worksapce-entitlements.md

```diff
@@ -0,0 +1,1004 @@
+    ClusterPath logicalcluster.Path `json:"clusterPath"`
+}
+
+// Entitlement describes an entitlement
+type Entitlement struct {
+    // ClusterName is the name of the logical cluster from which the entitlement is provided.
+    // NOTE: the entitlement review API is segmented by this cluster name in order to allow
+    // fine-grained authorization for the review API, so this field is superfluous; however, 
+    // duplicating the data here allows for the entitlement review object to be self-sufficient.
+    // +optional
+    ClusterName logicalcluster.Name `json:"clusterName"`
+
+    // The type of entitlement being described below is free-form data to be interpreted by
+    // the implementor of the review server. This type metadata allows for the object to be
+    // implemented as a discriminated union.
+    metav1.TypeMeta `json:",inline"`
+
+    // Spec is the free-form entitlement to be reviewed.
+    Spec runtime.RawExtension `json:"spec,omitempty"`
+}
+
+// EntitlementReviewStatus contains the result of the entitlement review.
+type EntitlementReviewStatus struct {
+    // Entitled is required. True if the entitlement is granted, false otherwise.
+    Entitled bool `json:"entitled"`
+    // Disqualified is optional. True if the entitlement  is not granted, otherwise
+    // false. If both entitled is false and disqualified is false, then the
+    // reviewer has no opinion on whether to grant the entitlement. Disqualified
+    // may not be true if entitled is true.
+    // +optional
+    Disqualified bool `json:"disqualified,omitempty"`
+    // Reason is optional. It indicates why a request was entitled or disqualified.
+    // +optional
+    Reason string `json:"reason,omitempty"`
+    // EvaluationError is an indication that some error occurred during the entitlement review.
+    // It is entirely possible to get an error and be able to continue determine entitlement status in spite of it.
+    // +optional
+    EvaluationError string `json:"evaluationError,omitempty"`
+}
+```
+
+The entitlement review API will be partitioned by workspace; entitlement reviews will be issued
+to the workspace-specific endpoint relevant to the entitlement and access to this API will be 
+guarded by Kubernetes RBAC in the workspace:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: core:entitlements:review
+rules:
+- apiGroups: ["core.kcp.io"]
+  resources:
+  - "entitlementreviews"
+  verbs: ["create"]
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: core:entitlements:review
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: core:entitlements:review
+subjects:
+- apiGroup: rbac.authorization.k8s.io
+  kind: Group
+  name: system:authenticated
+```
+
+Use of the entitlement review API might look like
+
+```
+> POST /services/entitlementreview/clusters/<entitlement-provider-cluster>/apis/core.kcp.io/v1alpha1/entitlementreviews
```

Comment:
this surprises. Would have expected to have the review as a local request, not against the provider cluster. Also having this under `/services` is odd.

Reply:
This is related to our private discussion in Slack about inversion of the mechanism:
- make entitlements properties of the acting workspace.
- allow a provider workspace to control entitlements.

Reply:
It's against the (local) view server, which is as local as we can get, right? It's not "against" the provider cluster, but we need to indicate somehow that we're reviewing an entitlement _from_ somewhere.

Reply:
why not `/clusters/<workspace-checked-for-entitlement>/...`?

Reply:
Commented somewhere else already: every service URL must be annonuced somewhere in another API object the caller has access to. `/services` as a prefix is not a convention and can vary by kcp installation and be completely different or on another host.